Add missing release notes

straight-shoota opened the pull request:
I've created blog posts for the missing releases 0.20.1 through 0.24.1. I extracted what seemed to be the most exciting features from each release to list some changes from each release but not copy the entire changelog.

Fixes #58

bew commented on _posts/2016-12-06-crystal-0.20.1-released.md:
```diff
@@ -0,0 +1,23 @@
+* `p` and `pp` use pretty printing.
+* `finished` macro hook places it's contents at the end of the program when all other types and methods are known.
+* Local variables can now have type restrictions.
+* `method_missing` hook can nowe define a method, useful for specifiying block arguments.
```
nowe => now

bew commented on _posts/2016-12-24-crystal-0.20.3-released.md:
```diff
@@ -0,0 +1,25 @@
+* `IO#gets`, `IO#each_line`, `String#lines`, `String#each_line`, etc. now chomp lines by default. You can pass `chomp: false` to prevent automatic chomping.
+* `HTTP::Handler` is no longer a class but a module.
+* Free variables now must be specified with `forall`, a single uppercase letter will not work anymore.
+* `libs` directory is removed from `CRYSTAL_PATH`, libraries should be put in `lib`.
```
Maybe just:
``The `libs` directory has been renamed to `lib`. Libraries must be moved to this new directory.``

straight-shoota replied:
I think the switch from libs to lib was already before, this change just removed the legacy option.

bew commented on _posts/2017-01-07-crystal-0.20.4-released.md:
```diff
@@ -0,0 +1,26 @@
+
+# Exciting Changes
+
+* `YAML::Emitter` was renamed to `JSON::Builder` and can used for any JSON serialization using `JSON.build`.
```
`JSON::Emitter` ? :smile:

bew commented on _posts/2017-01-22-crystal-0.20.5-released.md:
```diff
@@ -0,0 +1,23 @@
+* Improved performance for `String#index` and `Crypto::Bcrypt`.
+* `expect_raises` returns raised and rescued exception.
+* `Zip::File`, `Zip::Reader` and `Zip::Writer` are native readers for zip files that delegate compression to existing zlib module.
+* `crystal tool expand` expands macro call in a given location.
```
`Expands and prints` ?

bew commented on _posts/2017-04-01-crystal-0.22.0-released.md:
```diff
@@ -0,0 +1,24 @@
+
+* Backtrace function names are read from DWARF sections.
+* Sigfault and exceptions are printed to STDERR.
+* `PP` and `p` return the printed value.
```
`pp` (lowercase)

bew commented on _posts/2017-04-01-crystal-0.22.0-released.md:
```diff
@@ -0,0 +1,24 @@
+* Sigfault and exceptions are printed to STDERR.
+* `PP` and `p` return the printed value.
+* `XML::Node` supports setting and deleting attributes.
+* The `skip` macro method allows to skip a file: `{{ "{%" }} skip unless flag(:unix) %}`. No need to wrap entire files in `{{ "{%" }} if %}...{{ "{%" }} else %}` anymore!
```
What are the `{{` `}}` for? (Same below)

straight-shoota replied:
`{% ... %}` would be interpreted as Jekyll/Liquid template so it's just escaping that.

bew replied:
Oh ok

bew commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -0,0 +1,43 @@
+
+# Exciting Changes
+
+* With `ThinLTO` LLVM 4.0 and above allows faster release builds.
```
Add comma before `LLVM` ?

bew commented on _posts/2018-01-08-top-5-reasons-for-ruby-ists-to-use-crystal.md:
```diff
@@ -1,185 +1,185 @@
----
```
Why is there changes here?

straight-shoota replied:
Replaced CRLF with LF. It's an independent commit but I didn't feel like opening another PR for this.

RX14 commented on _posts/2016-12-06-crystal-0.20.1-released.md:
```diff
@@ -0,0 +1,23 @@
+# Exciting Changes
+
+* `p` and `pp` use pretty printing.
+* `finished` macro hook places it's contents at the end of the program when all other types and methods are known.
```
I assume `finished` was added, but this sentance doesn't make it clear that it was added, it makes it sound more like it's behaviour was changed.

RX14 commented on _posts/2016-12-06-crystal-0.20.1-released.md:
```diff
@@ -0,0 +1,23 @@
+* `p` and `pp` use pretty printing.
+* `finished` macro hook places it's contents at the end of the program when all other types and methods are known.
+* Local variables can now have type restrictions.
+* `method_missing` hook can now define a method, useful for specifiying block arguments.
```
This is a bit ambiguous on what it means, a link to the issue, or an example, would be ideal.

RX14 commented on _posts/2016-12-24-crystal-0.20.3-released.md:
```diff
@@ -0,0 +1,25 @@
+# Exciting Changes
+
+* `IO#gets`, `IO#each_line`, `String#lines`, `String#each_line`, etc. now chomp lines by default. You can pass `chomp: false` to prevent automatic chomping.
+* `HTTP::Handler` is no longer a class but a module.
```
Breaking changes should be noted as breaking.

straight-shoota replied:
There are a lot of them... I don't think we need that here.  I didn't even include all breaking changes in the blog posts, so what's the point in highlighting those that are mentioned?

mverzilli replied:
I think breaking changes should always be mentioned everywhere we published release notes. Other than that the PR is great!

RX14 commented on _posts/2017-01-22-crystal-0.20.5-released.md:
```diff
@@ -0,0 +1,23 @@
+
+* Improved performance for `String#index` and `Crypto::Bcrypt`.
+* `expect_raises` returns raised and rescued exception.
+* `Zip::File`, `Zip::Reader` and `Zip::Writer` are native readers for zip files that delegate compression to existing zlib module.
```
Again, that sounds like a fact not a change.

RX14 commented on _posts/2017-06-29-crystal-0.23.0-released.md:
```diff
@@ -0,0 +1,28 @@
+* The release is built with LLVM 3.8, so it's no longer supported by Debian 7 and CentOS.
+* Severity levels are now passed as `Severity` type to `Logger#formatter´.
+* `IO.select` was removed.
+* `Crystal::System::Random`
```
??

RX14 commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -0,0 +1,43 @@
+* `Time::Format` more robustly supports parsing sub-second precision.
+* `Tempfile` supports custom extensions.
+* The docs generator outputs documentation as JSON which can be easily parsed and used for external applications. The output directory of the docs generator can be configured using `--output` flag.
+* Garbage collection can be disabled using compiler flag `-Dgc_none`.
```
This is an implementation detail, and shouldn't be in these notes. It's not actually useful to anyone. (except @ysbaddaden as he implements a custom GC for crystal).

ysbaddaden replied:
I disagree, removing the GC is useful. For example a small lived programs may not care about leaking, since they'll most probably exit before the collector even runs.

RX14 replied:
I think people will get the wrong impression. I've already seen people interpret "no gc" be "like rust somehow magically". GC is magic to a lot of people, I don't think advertising this is a good idea.

RX14 commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -0,0 +1,43 @@
+* `HTML.escape` only escapes HTML special characters `&<>"'`. `HTML.escape` recognizes all HTML5 named entities.
+* Multiple assignment can no longer be used to create an array (`a = 1, 2, 3`).
+* `skip` macro method was renamed to `skip_file`.
+* Packages are built with LLVM 3.9.1 which should hopefully fix issues with LLVM error when compiling to release mode.
```
I'm pretty sure they're built with LLVM 4.

straight-shoota replied:
Ah yes, that was from 0.24.0. I merged them together because I don't think we need a blog post for the pre-release (at least not after the real release).

RX14 commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -0,0 +1,43 @@
+* Multiple assignment can no longer be used to create an array (`a = 1, 2, 3`).
+* `skip` macro method was renamed to `skip_file`.
+* Packages are built with LLVM 3.9.1 which should hopefully fix issues with LLVM error when compiling to release mode.
+* `rescue`, `ensure` and `else` blocks can be used directly in `do ...end` blocks.
```
Code example here would do wonders.

RX14 commented on _posts/2016-12-06-crystal-0.20.1-released.md:
```diff
@@ -0,0 +1,65 @@
+    id: Int32
+  }
+
+  macro attributes(**values)
```
`extra_attributes`

also this example seems a tad convoluted, not sure an example works well for this entry.

straight-shoota replied:
Well it certainly helps to be illustrated but I'm not so confident with that example myself.

RX14 commented on _posts/2016-12-06-crystal-0.20.1-released.md:
```diff
@@ -0,0 +1,65 @@
+
+# Exciting Changes
+
+* `p` and `pp` use pretty printing.
```
>now use

RX14 commented on _posts/2016-12-24-crystal-0.20.3-released.md:
```diff
@@ -0,0 +1,25 @@
+* `HTTP::Handler` is no longer a class but a module.
+* Free variables now must be specified with `forall`, a single uppercase letter will not work anymore.
+* `libs` directory is removed from `CRYSTAL_PATH`, libraries should be put in `lib`.
+* `class_getter`, `class_setter` and `class_property` macros where added.
```
>were

RX14 commented on _posts/2017-02-21-crystal-0.21.0-released.md:
```diff
@@ -0,0 +1,25 @@
+
+# Exciting Changes
+
+* Spec uses `it { ... }` instead of `assert { ... }`
```
>now uses

RX14 commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -0,0 +1,51 @@
+
+# Exciting Changes
+
+* With `ThinLTO`, LLVM 4.0 and above allows faster release builds.
```
>Added ThinLTO support for LLVM 4 and above, which speeds up release builds. You can use it by ...

RX14 commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -0,0 +1,51 @@
+
+* With `ThinLTO`, LLVM 4.0 and above allows faster release builds.
+* `SecureRandom` is renamed to `Random::Secure` and `SecureRandom.uuid` is replaced by a new type: `UUID` created using `UUID.random`
+* `BigDecimal` offers arbitrary precision, exact decimal numbers, supplementing `BigInt` and `BigFloat` which all are required by `require "big"`.
```
>BigDecimal was added, it offers ... . Supplementing ...

leave out the bit about require big.

RX14 commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -0,0 +1,51 @@
+# Exciting Changes
+
+* With `ThinLTO`, LLVM 4.0 and above allows faster release builds.
+* `SecureRandom` is renamed to `Random::Secure` and `SecureRandom.uuid` is replaced by a new type: `UUID` created using `UUID.random`
```
>a new UUID type, created

RX14 commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -0,0 +1,51 @@
+* `BigDecimal` offers arbitrary precision, exact decimal numbers, supplementing `BigInt` and `BigFloat` which all are required by `require "big"`.
+* `Set` instances can be used as condition for `case`, matching when the case variable is included in the set.
+* `Time::Format` more robustly supports parsing sub-second precision.
+* `Tempfile` supports custom extensions.
```
>now supports

RX14 commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -0,0 +1,51 @@
+* With `ThinLTO`, LLVM 4.0 and above allows faster release builds.
+* `SecureRandom` is renamed to `Random::Secure` and `SecureRandom.uuid` is replaced by a new type: `UUID` created using `UUID.random`
+* `BigDecimal` offers arbitrary precision, exact decimal numbers, supplementing `BigInt` and `BigFloat` which all are required by `require "big"`.
+* `Set` instances can be used as condition for `case`, matching when the case variable is included in the set.
```
>can now be

RX14 commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -0,0 +1,51 @@
+end
+{% endhighlight ruby %}</div>
+
+* `fun` names and lib function calls can start with uppercase.
```
>can now

RX14 commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -0,0 +1,51 @@
+{% endhighlight ruby %}</div>
+
+* `fun` names and lib function calls can start with uppercase.
+* YAML 1.1 core schema is fully supported.
```
>is now

RX14 commented on _posts/2017-12-23-crystal-0.24.1-released.md:
```diff
@@ -14,12 +14,12 @@ As any release it includes numerous bugfixes, cool features and performance impr
-* `Set` instances can be used as condition for `case`, matching when the case variable is included in the set.
+* Added `ThinLTO` support for LLVM 4.0 and above, which speeds up release builds. It can be activated using the compiler option `--lto=thin`.
+* `SecureRandom` is renamed to `Random::Secure` and `SecureRandom.uuid` is replaced by a new `UUID` type, created using `UUID.random`.
+* `BigDecimal` was added, it offers arbitrary precision, exact decimal numbers. Supplementing `BigInt` and `BigFloat` whith all three being included using `require "big"`.
```
>with

RX14 commented:
@straight-shoota ping? Can we get this merged please, it's fairly important.

straight-shoota commented:
I've fixed the typo, added links to the milestones and as @bcardiff and @RX14 suggested, marked all breaking changes and added breaking changes that were previously not mentioned.

matiasgarciaisaia commented:
Thank you *so much* @straight-shoota.

I've just merged this - the missing release notes are missing no more! 🎉
